app/controllers/api/storage.php

```diff
@@ -1109,17 +1109,32 @@
         }
 
         if ($size > APP_STORAGE_READ_BUFFER) {
-            for ($i = 0; $i < ceil($size / MAX_OUTPUT_CHUNK_SIZE); $i++) {
-                $response->chunk(
-                    $deviceFiles->read(
-                        $path,
-                        ($i * MAX_OUTPUT_CHUNK_SIZE),
-                        min(MAX_OUTPUT_CHUNK_SIZE, $size - ($i * MAX_OUTPUT_CHUNK_SIZE))
-                    ),
-                    (($i + 1) * MAX_OUTPUT_CHUNK_SIZE) >= $size
-                );
+            $buffer = '';
+            $bytesRead = 0;
+
+            while ($bytesRead < $size) {
+                // Determine how many bytes to read in this iteration
+                $bytesToRead = min(APP_STORAGE_READ_BUFFER, $size - $bytesRead);
+
+                // Read data from the device and append to the buffer
+                $buffer .= $deviceFiles->read($path, $bytesRead, $bytesToRead);
```

Comment:
I don't think we should append to buffer, because this means by the end whole file in the memory, and this logic is to prevent exactly that. Just keep the recent chunk in the buffer

Reply:
We are not appending the entire file to the buffer. If you look at the loop below, we remove each chunk from the buffer after processing. 

```
// Remove the sent chunk from the buffer
$buffer = substr($buffer, MAX_OUTPUT_CHUNK_SIZE);
```

Reply:
Why do this, why not instead of `$buffer .= $deviceFiles->read($path, $bytesRead, $bytesToRead);` simply use `$buffer = $deviceFiles->read($path, $bytesRead, $bytesToRead);` and when loop ends simply `$buffer = null`. The `substr` adds more processing